src/Functions/FunctionFile.cpp

```diff
@@ -47,80 +47,50 @@ class FunctionFile : public IFunction, WithContext
     {
         const ColumnPtr column = arguments[0].column;
-        const ColumnString * expected = checkAndGetColumn<ColumnString>(column.get());
-        if (!expected)
+        const ColumnString * column_src = checkAndGetColumn<ColumnString>(column.get());
+        if (!column_src)
             throw Exception(
                 fmt::format("Illegal column {} of argument of function {}", arguments[0].column->getName(), getName()),
                 ErrorCodes::ILLEGAL_COLUMN);
 
-        const ColumnString::Chars & chars = expected->getChars();
-        const ColumnString::Offsets & offsets = expected->getOffsets();
-
-        std::vector<String> checked_filenames(input_rows_count);
-
         auto result = ColumnString::create();
         auto & res_chars = result->getChars();
         auto & res_offsets = result->getOffsets();
 
         res_offsets.resize(input_rows_count);
 
-        size_t source_offset = 0;
-        size_t result_offset = 0;
+        fs::path user_files_absolute_path = fs::canonical(fs::path(getContext()->getUserFilesPath()));
+        std::string user_files_absolute_path_string = user_files_absolute_path.string();
+
+        // If run in Local mode, no need for path checking.
+        bool need_check = getContext()->getApplicationType() != Context::ApplicationType::LOCAL;
+
         for (size_t row = 0; row < input_rows_count; ++row)
         {
-            const char * filename = reinterpret_cast<const char *>(&chars[source_offset]);
+            StringRef filename = column_src->getDataAt(row);
+            fs::path file_path(filename.data, filename.data + filename.size);
 
-            fs::path user_files_absolute_path = fs::canonical(fs::path(getContext()->getUserFilesPath()));
-            fs::path file_path(filename);
             if (file_path.is_relative())
                 file_path = user_files_absolute_path / file_path;
-            fs::path file_absolute_path = fs::canonical(file_path);
-            checkReadIsAllowedOrThrow(user_files_absolute_path.string(), file_absolute_path);
-
-            checked_filenames[row] = file_absolute_path.string();
-
-            if (!fs::exists(file_absolute_path))
-                throw Exception(fmt::format("File {} doesn't exist.", file_absolute_path.string()), ErrorCodes::FILE_DOESNT_EXIST);
 
-            const auto current_file_size = fs::file_size(file_absolute_path);
+            /// Do not use fs::canonical or fs::weakly_canonical.
+            /// Otherwise it will not allow to work with symlinks in `user_files_path` directory.
+            file_path = fs::absolute(file_path).lexically_normal();
 
-            result_offset += current_file_size + 1;
-            res_offsets[row] = result_offset;
-            source_offset = offsets[row];
-        }
-
-        res_chars.resize(result_offset);
+            if (need_check && file_path.string().find(user_files_absolute_path_string) != 0)
```

Comment:
Consider using `fileOrSymlinkPathStartsWith(...)` from `filesystemHelpers.h`

Reply:
Ok, I will merge this, and make it in the next PR.

Btw, `fileOrSymlinkPathStartsWith` is not used in `file` table function.

Reply:
AFAIK, `TableFunctionFile` does not check path at all, because `StorageFile` [does](https://github.com/ClickHouse/ClickHouse/blob/051dddd8df72d4e8febac5ab8c266260a2dbf204/src/Storages/StorageFile.cpp#L131) (using `fileOrSymlinkPathStartsWith`)